src/polyfills/intersection-observer.js

```diff
@@ -19,14 +19,27 @@
  * See https://developer.mozilla.org/en-US/docs/Web/API/IntersectionObserver.
  */
 
-import {IntersectionObserverStub} from '../polyfillstub/intersection-observer-stub';
+import {
+  IntersectionObserverStub,
+  supportsDocumentRoot,
+} from '../polyfillstub/intersection-observer-stub';
 
 /**
  * @param {!Window} win
  */
 export function install(win) {
-  if (!win.IntersectionObserver) {
+  const NativeIntersectionObserver = win.IntersectionObserver;
+  if (!NativeIntersectionObserver) {
     win.IntersectionObserver = /** @type {typeof IntersectionObserver} */ (IntersectionObserverStub);
+  } else if (!supportsDocumentRoot(win)) {
+    win.IntersectionObserver = function (ioCallback, opts) {
+      if (opts && opts.root && opts.root.nodeType === 9) {
```

Comment:
Should we also check if it's in an iframe? E.g. opts.root === window.document && !isIframed(window) is mostly equivalent to the opts.root == null.

Reply:
I'm not sure I understand this comment. We have similar logic happening in the consumer (resources-impl) which requests for `root:document` if iframed and `root:null` otherwise. Feels too intrusive to put it here IMO, since implicit root and document root have some other subtle differences as we've learned

Reply:
Yeah, I see your point. Something like this we can leave to the caller. Agreed.

Reply:
Oh! I see one benefit to this check.
In a non iframed environment we can decide not to load the polyfill at all

Reply:
> In a non iframed environment we can decide not to load the polyfill at all.

There technically isn't a guarantee that even in a non-iframed (origin) environment some component won't eventually try tracking an iframe's document, so I don't think this optimization is valid

Reply:
Yep. I don't think we can optimize it this way here. It has to be a runtime decision on which one to instrument, so I think the current pattern should stay.